### PR TITLE
Moving GCM endpoint over to using web push protocol

### PR DESCRIPTION
**Pull request by gauntface**
This removes the GCM protocol use. It does leave in the Authorization header needed for the GCM API Key though.

### CODE REVIEW

**Comment by gauntface** [comment]
Fixes #174 and #125 hopefully

File: index.js [hunk 1]
```diff
@@ -143,28 +141,6 @@ function sendNotification(endpoint, params) {
...
-        }
-        requestPayload = JSON.stringify(gcmObj);
-
-        options.path = options.path.substring(0, options.path.length - subscriptionId.length - 1);
-
-        options.headers['Authorization'] = 'key=' + gcmAPIKey;
-        options.headers['Content-Type'] = 'application/json';
-      }
-
       if (vapid && !isGCM) {
```
Comment (marco-c) [inline index.js]: isGCM is used before its definition here.

File: index.js [hunk 2]
```diff
@@ -188,7 +164,7 @@ function sendNotification(endpoint, params) {
         options.headers['Authorization'] = 'Bearer ' + jwt;
         var key = 'p256ecdsa=' + urlBase64.encode(vapid.publicKey);
         if (options.headers['Crypto-Key']) {
-          options.headers['Crypto-Key'] += ',' + key;
+          options.headers['Crypto-Key'] += ';' + key;
```
Comment (marco-c) [inline index.js]: Shouldn't this be a comma?
At least this is what Autopush is expecting: https://github.com/mozilla-services/autopush/blob/f546ed78585976d33534595dafcef401c118ea43/autopush/crypto_key.py and https://github.com/mozilla-services/autopush/blob/4361bbe07ee4b962622a3c16ddf632405a599158/autopush/tests/test_cryptokey.py

Comment (gauntface) [inline index.js]: If I put in a comma with GCM, it actually broke the GCM API and discussing this with @beverloo he pointed out:

 > Semi-colon separates fields in a single header value, comma separates header values.

which I cant find in a spec anywhere for semicolon but regarding comma [from this spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2):

> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma. The order in which header fields with the same field-name are received is therefore significant to the interpretation of the combined field value, and thus a proxy MUST NOT change the order of these field values when a message is forwarded.

This *might* have been one of the issues I was having with VAPID support on Chrome.

Comment (marco-c) [inline index.js]: Yes, semicolons separate parameters, commas separate values.

From http://httpwg.org/http-extensions/draft-ietf-httpbis-encryption-encoding.html#crypto-key:
> The Crypto-Key header field uses the extended ABNF syntax defined in Section 1.2 of [RFC7230]

From https://tools.ietf.org/html/rfc7230#section-1.2:
> This specification uses the Augmented Backus-Naur Form (ABNF) notation of [RFC5234] with a list extension, defined in Section 7, that allows for compact definition of comma-separated lists [...]

Then, in https://tools.ietf.org/html/draft-ietf-webpush-vapid-00#section-4:
> Note that with push message encryption [I-D.ietf-webpush-encryption], this results in two values in the Crypto-Key header field, one with the a "p256dh" key and another with a "p256ecdsa" key.

So, in this case, I think it should be a comma.

Comment (gauntface) [inline index.js]: My only counter argument is:

> one with the a "p256dh" key and another with a "p256ecdsa" key

They are are two (key, value) pairs which I'd view as two parameters. This is confused by:

>  this results in two values

Seems like a spec ambiguity @beverloo + @martinthomson

Comment (beverloo) [inline index.js]: /cc @martinthomson 

Which would, in turn, violate [draft-ietf-webpush-encryption](https://tools.ietf.org/html/draft-ietf-webpush-encryption-02#section-3.2):

> An application server MUST include exactly one entry in each of the Encryption and Crypto-Key header fields.

Semantically, having them in different entries makes sense because `p256ecdsa` has a different intended consumer (the push service) than `dh` (the user agent). This is unfortunately not compatible with all of today's implementations. FCM and Chrome enforce the single-entry restriction set by webpush-encryption, and will fail to deliver the message.

tl;dr— I suggest using `;` for compatibility reasons until we [fix the spec](https://github.com/webpush-wg/webpush-encryption/issues/6) and know the implementations can deal with this.

Comment (marco-c) [inline index.js]: From the implementation of `Autopush` it looks like, even if it assumed different values to be separated by comma, it doesn't break if you use a semicolon instead.
@jrconlin tested and confirmed this, so I think we can use the semicolon for the time being.

**Comment by martinthomson** [comment]
FWIW, comma was the intent.  The semi-colon is probably fine.
